Squash feature/PBST-0001-initial-branch

j-stahl opened the pull request:
## Contents
Complete CLI solution for profitbricks to create snapshots based on server names. 

**Squash message**
```
Implement symfony commands, readme, license, classes for object properties and the api connectivity.
#DLR-??? Fixed
```

j-stahl committed:
Initial Commit

Contains first version.

j-stahl committed:
Change headings.

Change headings.

mriegert commented on commands/snapshotautocreatecommand.php:
```diff
@@ -0,0 +1,117 @@
+        if (is_readable('config.ini'))
+        {
+            $this->config = parse_ini_file('config.ini', true);
+            if (!isset($this->config['api']['user']) || !isset($this->config['api']['password']))throw new Exception("No user or no password configured to connect ProfitBricks!");
```
This code could run into PHP7-problems with the variable handling order changes.
If it comes to invalid array-to-string-conversions, there have to be added some curly braces:

for example:` isset($this->{config['api']['user']})`  reverts the code to the intended behaviour

mriegert commented on commands/snapshotautocreatecommand.php:
```diff
@@ -0,0 +1,117 @@
+                        if ($latestDiskBackup<$latestServerBackup) $latestServerBackup = $latestDiskBackup;
+                        if ($latestDiskBackupByScript<$latestServerBackupByScript) $latestServerBackupByScript = $latestDiskBackupByScript;
+                    }
+                    $tableRows[] = array ($dataCenter->name, $virtualMachine->name, "YES  ".$this->config[$virtualMachine->name]["snapshotInterval"]." days", $latestServerBackup->format("d.m.Y"), $latestServerBackupByScript->format("d.m.Y"));
```
Is this "YES" intended to be there?

mriegert commented on commands/snapshotautodeletecommand.php:
```diff
@@ -0,0 +1,136 @@
+        }
+        else throw new Exception("Error during reading config.ini!");
+        $this->profitBricksApi->setUserName($this->config["api"]["user"]);
+        $this->profitBricksApi->setPassword($this->config["api"]["password"]);
```
**It is no review request** but as I look for code duplication (as usual) and read this code block in each command, I am thinking about refactoring the marked block into an init-method of a (to be created) baseclass for all the commands.

mriegert commented on commands/snapshotautodeletecommand.php:
```diff
@@ -0,0 +1,136 @@
+                    }
+                    else  $tableRows[] = array ($snapshot->id, $snapshot->name, $snapshot->description, $snapshot->createdDate->format("d.m.Y H:i"), $snapshot->size." GB", "no configuration for server ".$virtualMachine->name."!");
+                }
+                else  $tableRows[] = array ($snapshot->id, $snapshot->name, $snapshot->description, $snapshot->createdDate->format("d.m.Y H:i"), $snapshot->size." GB", "no valid server name!");
```
This bunch of else-cases also share a lot of redundant parts. A possibility of reducing redundancy here could be to only set the changing part (last array index) in the else-cases and finish the operation by setting:
`$tableRows[] = array ($snapshot->id, $snapshot->name, $snapshot->description, $snapshot->createdDate->format("d.m.Y H:i"), $snapshot->size." GB", $SET_BY_ELSE_CASE);`

mriegert commented on commands/snapshotautodeletecommand.php:
```diff
@@ -0,0 +1,136 @@
+        else throw new Exception("There are no snapshots at ProfitBricks!");
+    }
+
+    private function getVirtualMachineFor(Snapshot $snapshot)
```
This method needs a doc block, especially to let an IDE know what is returning.

mriegert commented on commands/snapshotautodeletecommand.php:
```diff
@@ -0,0 +1,136 @@
+    {
+        foreach ($this->virtualMachines as $virtualMachine)
+        {
+            if (strpos($snapshot->name, $virtualMachine->name)!==false)// && strpos($snapshot->description, $virtualMachine->name)!==false)
```
The commented expression should be deleted.

mriegert commented on commands/snapshotcheckercommand.php:
```diff
@@ -0,0 +1,119 @@
+        $profitBricksApi = new ProfitBricksApi();
+        $profitBricksApi->setUserName($this->config["api"]["user"]);
+        $profitBricksApi->setPassword($this->config["api"]["password"]);
+        /** @var Snapshot[] $snapshots */
```
This documentation is not wrong but it would better to document the called method ProfitBricksApi::snapshots() and define the return value by @return Snapshot[]. Then there would be no warnings from our IDE.

mriegert commented on commands/snapshotcheckercommand.php:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @param $_virtualMachineId
```
Please add at least the types of the input parameters.
So no one has to look up in the code that for example $_backupDates cannot be an array.

mriegert commented on commands/snapshotcheckercommand.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @param ProfitBricksApi $_profitBricksApi
+     */
+    private function readVirtualMachines(ProfitBricksApi &$_profitBricksApi)
```
The `&`-operator should be removed here and avoided whenever it is possible.

mriegert commented on commands/snapshotcreatefromattacheddiskscommand.php:
```diff
@@ -0,0 +1,104 @@
+        $io->title("Snapshot Creation Tool");
+        $tableRows = array();
+        foreach ($dataCenters as $dataCenter)
+        {   /** @var VirtualMachine $virtualMachine */
```
As function dataCenters(), also function virtualMachines(DataCenter) should define in its method documentation to return the correct object types. Then this loop would need no extra typing to get rid off some warnings from the IDE.

mriegert commented on commands/snapshotcheckercommand.php:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @param ProfitBricksApi $_profitBricksApi
+     */
+    private function readVirtualMachines(ProfitBricksApi &$_profitBricksApi)
```
What about moving this function into the ProfitBricksApi?

j-stahl replied:
good idea

mriegert commented on commands/snapshotcreatefromattacheddiskscommand.php:
```diff
@@ -0,0 +1,104 @@
+                if (count($input->getArgument("server"))>0)
+                {
+                    foreach ($input->getArgument("server") as $server)
+                    {
```
Add comment: loop over all VirtualMachines and do the operation for selected servers
(maybe in 2 separat lines)

mriegert commented on commands/snapshotcreatefromattacheddiskscommand.php:
```diff
@@ -0,0 +1,104 @@
+                    }
+                }
+                else
+                {
```
add short comment that the operation will be done for all available VirtualDisks

mriegert commented on commands/snapshotdeletecommand.php:
```diff
@@ -0,0 +1,143 @@
+    public function __construct()
+    {
+        parent::__construct();
+        $this->before = new \DateTime();
```
The command option can be let **before** but to make the code more clear the variable name here should be `deleteBeforeTimestamp`.

mriegert commented on commands/snapshotdeletecommand.php:
```diff
@@ -0,0 +1,143 @@
+ * Class SnapshotDeleteCommand
+ *
+ * This command deletes snapshot(s) at ProfitBricks by snapshot ID or filtered by date (deletes older than).
+ */
```
As discussed, this command needs a check of its workflow to assert that no one deletes all snapshots in an unindented way.
Method interact() seems to be here left over from a solution try.

mriegert commented on commands/snapshotdeletecommand.php:
```diff
@@ -0,0 +1,143 @@
+            ->addOption("before","b", InputOption::VALUE_REQUIRED, "lists all snapshot before this date time!", null);
+    }
+
+    protected function interact(InputInterface $input, OutputInterface $output)
```
Check if removable.

mriegert commented on commands/snapshotdeletecommand.php:
```diff
@@ -0,0 +1,143 @@
+        }
+        $sumSize = ceil($sumSize / 100);
+        $sumSize = $sumSize / 10;
+        $sumSize = str_replace('.',',',(string)$sumSize);
```
It is quite common to have a method like niceSize(int $_size) which returns a fully divided and human readable string.
If it comes to values from kilobytes to terabytes with multiple transformations, this would be needed here.

mriegert commented on composer.json:
```diff
@@ -0,0 +1,14 @@
+    "description": "ProfitBricks Snapshot Tool",
+    "type": "library",
+    "require": {
+        "symfony/console": "^3.1"
```
Do you have checked that this is the least possible version?

j-stahl replied:
I did not check exact which versions of symfony console commands are compatible and which are not.

mriegert commented on demo-config.ini:
```diff
@@ -0,0 +1,22 @@
+
+[mail.ducktales.com]
+snapshotStartDay = "friday"
+snapshotInterval = 21
```
As I did not directly understand this value **snapshotInterval** , I would like have a commentation.
commentation in .ini-files with `; COMMENT`

mriegert commented on pbst.php:
```diff
@@ -0,0 +1,33 @@
+ */
+
+require_once ("vendor/autoload.php");
+foreach (glob(getcwd()."/commands/*command.php") as $filename)
```
This glob in the current working dir only is not optimal here because a command should work independentely from the shell path where  it got executed.
It can be simply replaced by:  `foreach (glob(__DIR__."/commands/*command.php") as $filename) `

But we could provide an autoload-mechanism for the includes which is state-of-the-art and quickly added:
```
/**
 * Custom auto loader for own files.
 * @param string $_className The name of the class that gets sourced.
 */
function customAutoLoader($_className) {
    foreach (array(".", "commands", "profitbricksapi") as $directory) {
        $filepath= __DIR__."/".$directory."/".$_className.'.php';
        if (file_exists($filepath))
        {
            require_once filepath;
            break;
        }
    }
}
spl_autoload_register("customAutoLoader");
```
This autoloads requirements and let you remove each following require-line (imports by **use** are still needed).
So import your library: ```use ProfitBricksApi\ProfitBricksApi;``` and think about one new name.

mriegert commented on profitbricksapi/datacenter.php:
```diff
@@ -0,0 +1,89 @@
+namespace ProfitBricksApi;
+
+/**
+ * Class DataCenter describes the properties of a data center at ProfitBricks.
```
Here and for all other data-containers (snapshot, virtualdisk, virtualmachine), I would a comment that these are quite simple data-containers which wrap all available parameters and in a well formulated sentence,
that these here defined values can resist (and further used in the code) even if ProfitBricks changes something.
The only need would be to adapt the simple wrapper here.

mriegert commented on profitbricksapi/datacenter.php:
```diff
@@ -0,0 +1,89 @@
+ *  $dataCenter->id                             id
+ *  $dataCenter->type                           mostly 'snapshot'
+ *  $dataCenter->href                           link (maybe for download)
+ *  $dataCenter->metadata                   object
```
Maybe fix the indenting in all lines before publishing the tool?

mriegert commented on profitbricksapi/profitbricksapi.php:
```diff
@@ -0,0 +1,259 @@
+    public function snapshots()
+    {
+        $snapshots = false;
+        $profitBricksSnapshots = $this->readFromProfitBricks(self::profitBricksSnapshotApi);
```
As dicussed (and not a review request), I would prefer an API which returns empty arrays instead of bool|array.
This would reduce this method to this:

```
$snapshots = array();
foreach ($this->readFromProfitBricks(self::profitBricksSnapshotApi) as $snapshot)
{
     $snapshots[$snapshot->id] = new Snapshot($snapshot);
}
return $snapshots;
```

mriegert replied:
The code using this function here would also not need the IF-check.
Checking for empty arrays is still possible (to have the check if needed)

mriegert commented on profitbricksapi/profitbricksapi.php:
```diff
@@ -0,0 +1,259 @@
+    {
+        $virtualMachines = false;
+        $api = str_replace("[data-center-id]", $_dataCenter->id, self::profitBricksServerApi);
+        foreach ($this->readFromProfitBricks($api) as $virtualMachine)
```
Theoretically, it would need here this IF-check for the case that nothing usable gets returned.

mriegert commented on profitbricksapi/profitbricksapi.php:
```diff
@@ -0,0 +1,259 @@
+        $api = str_replace("[data-center-id]", $_dataCenterId, self::profitBricksDiskApi);
+        $api = str_replace("[server-id]", $_virtualMachine->id, $api);
+        $latestFullSnapshot = null;
+        foreach ($this->readFromProfitBricks($api) as $virtualDisk)
```
Theoretically, it would need here this IF-check for the case that nothing usable gets returned.

mriegert replied:
After reading the last (similar) functions, I would find it a good hint for further readers to name the iterator-variables $virtualDisk (+ virtualMachines + snapshots +...) $virtualDiskConfig and so on to signal that these only are JSON-configs returned by the server and not our wrapper objects.

mriegert commented on pbst.php:
```diff
@@ -0,0 +1,33 @@
+use Symfony\Component\Console\Application;
+
+$application = new Application();
+$application->add(new \PBST\Command\SnapshotListCommand());
```
I would like to add a bit coolness to the commands entrypoint and replace these add-operations with a loop:
```
$application = new Application();
foreach(array("SnapshotList", "SnapshotDelete", ...) as $command)
{
    $command = "\\PBST\\Command\\".$command;
    $application->add(new $command());
}
```

If something goes wrong with the autoloading stuff which might be not able to load the namespaced classes,
this replacement will always work.

[ As fallback-strategy, the require_once-repetitions can also replaced with a normal glob-loop ]

mriegert replied:
There exists a minor naming convention problem with the used namespaces and the directory structure:
\PBST\Command\SnapshotDeleteCommand() should be located in an additional subfolder PBST
like it is organized with the contained classes in the other subdirectory **ProfitBricksApi**
For example: `class ProfitBricksApi in \ProfitBricksApi\ProfitBricksApi`

mriegert commented on profitbricksapi/profitbricksapi.php:
```diff
@@ -0,0 +1,259 @@
+    const profitBricksDeleteSnapshot = "https://api.profitbricks.com/cloudapi/v3/snapshots/[snapshot-id]";
+    const profitBricksCreateSnapshot = "https://api.profitbricks.com/cloudapi/v3/datacenters/[data-center-id]/volumes/[volume-id]/create-snapshot";
+
+    public function __construct()
```
Empty constructor is not needed.

mriegert commented on profitbricksapi/profitbricksapi.php:
```diff
@@ -0,0 +1,259 @@
+
+    /**
+     * This function imports strings and JSON from ProfitBricks, which it cuts down to an usable array of objects.
+     *
```
Maybe add an advice that this method is quite important as it is able to retrieve nearly all available information about DataCenters and their maintained VirtualMachines.

I do not know if you can do arbitrary calls on the ProfitBricks API without costs (method is called again for each operation).

mriegert committed:
Add a possible baseclass implementation as proposal

Add a possible baseclass implementation as proposal

mriegert reviewed (changes requested):
First, even if I wrote so much, I liked the code and find it good readable. Many comments are tagged as NO REVIEW REQUEST and need not to be done.

The code wraps well the ProfitBricks cloud API.

I provided a baseclass which can reduce some redundancy and contains some documentation for further readers of the code.

I have proposed untested code inside some comments which can be used in the code or as info ;)

A namespace naming convention is a bit hurt.

j-stahl committed:
Fix review issues.

Fix review issues.

j-stahl committed:
Implement Base Class

Implement Base Class and fix path issue.

j-stahl committed:
Fix deletion bug.

Snapshot deletion did not work properly because of wrong code
modifications during installing the basecommand class.

j-stahl committed:
Revert "Fix deletion bug."

This reverts commit ef99b40e63546c8e259646317b1be2d6790fe05c.

j-stahl committed:
Fix deletion bug

Bug has arisen due to the wrong coding of the basecommand class.

mriegert commented on pbst.php:
```diff
@@ -0,0 +1,64 @@
+require_once ('profitbricksapi/virtualmachine.php');
+require_once ('profitbricksapi/virtualdisk.php');
+require_once ('profitbricksapi/snapshot.php');
+*/
```
The commented block should be removed.

mriegert commented on pbst.php:
```diff
@@ -0,0 +1,64 @@
+$application->add(new \PBST\Commands\SnapshotCheckerCommand());
+$application->add(new \PBST\Commands\ServerListCommand());
+$application->add(new \PBST\Commands\DiskListCommand());
+*/
```
same here

mriegert reviewed (changes requested):
Last Iteration!
There are two commented blocks left in pbst.php that should be removed. Then I can merge it.

j-stahl committed:
Removed unneeded commented code.
Removed unneeded commented code.

mriegert reviewed (approved):
Good work, I will merge it now.
